### PR TITLE
Large data counts support for MPI Communication

### PR DESCRIPTION
**Pull request by JuanPedroGHM**
## Due Diligence

- General:
    - [ ]  **title** of the PR is suitable to appear in the [Release Notes](https://github.com/helmholtz-analytics/heat/releases/latest)
- Implementation:
    - [ ] unit tests: all split configurations tested
    - [ ] unit tests: multiple dtypes tested
    - [ ] benchmarks: created for new functionality
    - [ ] benchmarks: performance improved or maintained
    - [ ] documentation updated where needed

## Description
Some MPI implementation are limited to sending only 2^31-1 elements at once. As far as I have tested, this also applies for OpenMPI 4.1 and 5.0, because support has not been added to mpi4py. (At least in my tests it failed).

This small changes uses the trick described [here](https://ieeexplore.ieee.org/document/7018162), to pack contiguous data into an MPI Vector, extending the limit of elements being sent. 

This is for contiguous data, as non-contiguous data is already packed in recursive vector data types, reducing the need to apply this trick. 

Issue/s resolved: #

## Changes proposed:

- MPI Vector to send more than 2^31-1 elements at once.
-
-
-

## Type of change
- Bug fix (non-breaking change which fixes an issue)

#### Does this change modify the behaviour of other functions? If so, which?
yes, probably a lot of them.

### CODE REVIEW

**Comment by mrfh92** [comment]
I have encountered the following problem: 

```
import heat as ht 
import torch

shape = (2 ** 10, 2 ** 10, 2 ** 11)

data = torch.ones(shape, dtype=torch.float32) * ht.MPI_WORLD.rank
ht.MPI_WORLD.Allreduce(ht.MPI.IN_PLACE, data, ht.MPI.SUM)
```
results in the following error: 
```
  File /heat/heat/core/communication.py", line 915, in Allreduce
    ret, sbuf, rbuf, buf = self.__reduce_like(self.handle.Allreduce, sendbuf, recvbuf, op)
  File "/heat/heat/core/communication.py", line 895, in __reduce_like
    return func(sendbuf, recvbuf, *args, **kwargs), sbuf, rbuf, buf
  File "src/mpi4py/MPI.src/Comm.pyx", line 1115, in mpi4py.MPI.Comm.Allreduce
mpi4py.MPI.Exception: MPI_ERR_OP: invalid reduce operation
```

With 2 ** 10 in the last entry of shape, there is not problem, so it seems to be related to large counts.

**Comment by JuanPedroGHM** [comment]
### Benchmarks results - Sponsored by [perun](https://github.com/Helmholtz-AI-Energy/perun)

| function            |   mpi_ranks | device   | metric   |     value |   ref_value |        std |   % change | type            | alert   |   lower_quantile |   upper_quantile |
|---------------------|-------------|----------|----------|-----------|-------------|------------|------------|-----------------|---------|------------------|------------------|
| concatenate         |           4 | GPU      | RUNTIME  | 0.0993419 |   0.0901117 | 0.0164166  |   10.243   | jump-detection  | True    |       nan        |       nan        |
| lanczos             |           4 | GPU      | RUNTIME  | 0.623046  |   0.61123   | 0.00828365 |    1.93324 | trend-deviation | True    |         0.594628 |         0.621807 |
| hierachical_svd_tol |           4 | GPU      | RUNTIME  | 0.12677   |   0.122009  | 0.0070573  |    3.9029  | trend-deviation | True    |         0.118723 |         0.124756 |

[Grafana Dashboard](http://57bc8d92-72f2-4869-accd-435ec06365cb.ka.bw-cloud-instance.org:3000/d/adjpqduq9r7k0a/heat-cb?orgId=1)
Last updated: 2025-01-27T11:03:45Z

**Comment by mrfh92** [comment]
Could there be the problem that for all communication involving MPI-Operations like MPI.SUM etc. such an operation is not well-defined on the MPI-Vector construction chosen for the buffers?

**Comment by JuanPedroGHM** [comment]
> Could there be the problem that for all communication involving MPI-Operations like MPI.SUM etc. such an operation is not well-defined on the MPI-Vector construction chosen for the buffers?

Have you found some bug? I don't think it should be an issue, as the vector datatype is just pointing to where the data is, where it needs to go, and it in what order. As long as both send and recv buffers are well-defined by the datatype, there should not be an issue with MPI operations.

**Comment by mrfh92** [comment]
The example with Allreduce I posted above caused an error for me.

**Comment by JuanPedroGHM** [comment]
Had a look, and found out that reduction operations with derived datatypes require custom made reduction functions to work. Latest commits handle that. It came with a slightly bigger than expected refactoring of `__reduce_like`, but it might have a lot of benefits. 

After some early and none conclusive tests, it might have a both a runtime and memory usage improvement. 

```python
import heat as ht 
import torch
import perun

from heat.core.communication import CUDA_AWARE_MPI
print(f"CUDA_AWARE_MPI: {CUDA_AWARE_MPI}")

world_size = ht.MPI_WORLD.size
rank = ht.MPI_WORLD.rank
print(f"{rank}/{world_size} reporting for duty!")
dtype = torch.int64

@perun.monitor()
def cpu_inplace_contiguous(n_elements, shape):
    data: torch.Tensor = torch.arange(n_elements,dtype=dtype).reshape(shape)
    ht.MPI_WORLD.Allreduce(ht.MPI.IN_PLACE, data, ht.MPI.SUM)
    del data

@perun.monitor()
def cpu_inplace_non_contiguous(n_elements, shape):
    data: torch.Tensor = torch.arange(n_elements,dtype=dtype).reshape(shape)
    data = data.permute(2,1,0)
    ht.MPI_WORLD.Allreduce(ht.MPI.IN_PLACE, data, ht.MPI.SUM)
    del data

@perun.monitor()
def cpu_contiguous(n_elements, shape):
    data: torch.Tensor = torch.arange(n_elements,dtype=dtype).reshape(shape)
    zeros = torch.zeros(shape, dtype=dtype)
    ht.MPI_WORLD.Allreduce(zeros, data, ht.MPI.SUM)
    del data
    del zeros

@perun.monitor()
def cpu_non_contiguous(n_elements, shape):
    data: torch.Tensor = torch.arange(n_elements,dtype=dtype).reshape(shape)
    data = data.permute(2,1,0)
    zeros = torch.zeros(shape, dtype=dtype)
    zeros = zeros.permute(2,1,0)
    ht.MPI_WORLD.Allreduce(data, zeros, ht.MPI.SUM)
    del data
    del zeros

@perun.monitor()
def gpu_inplace_contiguous(n_elements, shape):
    data: torch.Tensor = torch.arange(n_elements, dtype=dtype, device=f"cuda:{rank}").reshape(shape)
    print(f"Rank {rank} working on device {data.device}")
    ht.MPI_WORLD.Allreduce(ht.MPI.IN_PLACE, data, ht.MPI.SUM)
    del data

@perun.monitor()
def gpu_inplace_non_contiguous(n_elements, shape):
    data: torch.Tensor = torch.arange(n_elements, dtype=dtype,device=f"cuda:{rank}").reshape(shape)
    data = data.permute(2,1,0)
    ht.MPI_WORLD.Allreduce(ht.MPI.IN_PLACE, data, ht.MPI.SUM)
    del data

@perun.monitor()
def gpu_contiguous(n_elements, shape):
    data: torch.Tensor = torch.arange(n_elements, dtype=dtype,device=f"cuda:{rank}").reshape(shape)
    zeros = torch.zeros(shape, dtype=dtype,device="cuda")
    ht.MPI_WORLD.Allreduce(data, zeros, ht.MPI.SUM)
    del data
    del zeros

@perun.monitor()
def gpu_non_contiguous(n_elements, shape):
    data: torch.Tensor = torch.arange(n_elements, dtype=dtype,device=f"cuda:{rank}").reshape(shape)
    data = data.permute(2,1,0)
    zeros = torch.zeros(shape, dtype=dtype,device="cuda")
    zeros = zeros.permute(2,1,0)
    ht.MPI_WORLD.Allreduce(data, zeros, ht.MPI.SUM)
    del data
    del zeros

    

if __name__ == "__main__":
    base_shape = (500, 500, 500 * world_size)
    shape = base_shape
    n_elements = torch.prod(torch.tensor(shape))
    print(f"N Elements: {n_elements}, Memory: {n_elements * dtype.itemsize / (1000**3)}Gb")

    print("cpu_contigous")
    cpu_contiguous(n_elements, shape)
    print("cpu_non_contigous")
    cpu_non_contiguous(n_elements,shape)
    print("cpu_inplace_contiguous")
    cpu_inplace_contiguous(n_elements, shape)
    print("cpu_inplace_non_contiguous")
    cpu_inplace_non_contiguous(n_elements, shape)

    print("gpu_inplace_contiguous")
    gpu_inplace_contiguous(n_elements, shape)
    print("gpu_inplace_non_contiguous")
    gpu_inplace_non_contiguous(n_elements, shape)
    print("gpu_contigous")
    gpu_contiguous(n_elements, shape)
    print("gpu_non_contigous")
    gpu_non_contiguous(n_elements,shape)
```

Here is the script that I was using to test this. I'll try to get more conclusive results during the week. (And to fix the benchmarks, this is breaking them for some reason)

**Comment by JuanPedroGHM** [comment]
image
image
image

**Review by mrfh92 — APPROVED** [review]
As far as I can judge, these changes look fine. As testing this approach requires quite large messages, it won't be possible to test every functionality and/or exception properly within the CI; thus I'd suggest to ignore the non-100%-patch-coverage for this PR. 

Thanks for the work! :)

**Comment by mrfh92** [comment]
I am not sure what happens exactly, but there seem to be problems on the AMD-runner (it crashes in test_communication) and also for PyTorch 2.0.1 (and all Python versions)
